### ISSUE TITLE
Performance: Suboptimal repeat element lifecycle

### ISSUE THREAD

**Issue by niieani** [aurelia/templating-resources#213]
* [Example (with `@bindable`)](https://gist.run/?id=47985f620e3632b40f18197e52ecf77f)
* [Example (with `<compose>`)](https://gist.run/?id=bd1ee888a9311027156b29b7df6ac27e)

1. Open either of the examples
2. Open Developer Tools console to see when elements bind/unbind. 
3. Click the checkbox to filter the object with the value '8'.

What happens is that when one toggles the value filter, Aurelia doesn't diff between the unfiltered and filtered elements, but removes **all** of them from the view and recreates all of them again, each time. In the developer console you can see that upon enabling, each of the Components is unbound and bound again, only this time without the filtered value 8. The same thing happens when you toggle it back. This leads to very bad performance when there are more elements - try changing the number of elements from 10 to 1000 in the for loop of `app`. With 1000 elements my MBP takes at least 1-2 seconds to mutate the DOM, while all that should've happened is the removal or addition of a single Component - something that should be instantaneous.

The workaround is to never change the underlying array/map references, and in turn not to use value-converters for repeaters (instead - always mutate the original instance that the View binds to). But then ValueConverters, one of the highlight features of Aurelia, loose a lot of their appeal. A task like filtering an array is pretty common in any app, even the simplest one such as the TodoMVC. It would be great if Aurelia tackled this problem systemically, because manually mutating an array for use-cases such as filtration requires a lot of manual hackery (keeping another "reference array" and manually multi-splicing the "filtered array" for each filtration/de-filtration action - huge pain in the butt!).

Sometimes it is also easier for the developer to simply provide a fresh array/map instead of mutating the old one; this is especially true when using immutable programming styles or observables. It would be great if Aurelia could simply figure out which elements got removed and which got added, without the need for the developer to reimplement this for every array.

Personally, this is a deal-breaker for me, since competitor frameworks do not have this issue - React, Cycle.js and any VirtualDom implementation such as snabbdom will always diff the values in order to optimally mutate the DOM - what matters are the individual components, not the iterable structure that holds them together.

**Comment by EisenbergEffect** [comment]
@martingust @jdanyow Can we look into doing some more improvements in this area?
@niieani On this note, if you have a long list, you should really be using the ui-virtualization library so that you don't even make the browser render things that aren't' visible. Just install the plugin and switch your `repeat` to a `virtual-repeat`.

**Comment by niieani** [comment]
@EisenbergEffect I am aware of the ui-virtualization plugin. However, if the template rendered by the repeater is really complex (or nests many components), the performance problem will appear even if the list is only 10 items long, simply because of the tear-down and set-up of everything inside the repeat template involved (bindings, unbindings, composition, view-creation, etc, etc.). 

Whenever the instance of the observed Array/Map/Set changes, instead of resetting all of the rendered elements, as Aurelia currently does, I believe the solution would be to:
- diff the old instance and the new instance, finding which elements were added, which removed and the shift of their positions
- commit those changes, by:
  - unbinding/destroying the removed elements and removing them from the DOM
  - moving the pre-existent elements around in the DOM to match the new order (if different)
  - creating/binding new elements, and adding them to the DOM to the correct place

**Comment by EisenbergEffect** [comment]
I thought we were already doing this but maybe there's a bug or some
conflict with valued converters.

**Comment by niieani** [comment]
@EisenbergEffect Well it might be a bug, but it's definitely not a conflict with value converters, because the same problem is present when we manually change the instance of the Array/Map/Set from the ViewModel.
[Example (swapping arrays in the ViewModel)](https://gist.run/?id=123dc3b0b8353992ac93e46f4e44a487)

**Comment by martingust** [comment]
@niieani We are analyzing the iterated view [here](https://github.com/aurelia/templating-resources/blob/master/src/repeat.js#L67), if it doesn't require lifecycle the optimizations you suggested are happening [here](https://github.com/aurelia/templating-resources/blob/master/src/array-repeat-strategy.js#L43), if the view (i.e the compose element) do require full lifecycle all views will be removed. You can optimise this process by implemating view chaching by simply adding `view-cache="*"`- attribute on the ` compose` element.

**Comment by stsje** [comment]
I tried adding the `view-cache="*"` attribute in the examples, but not seeing it having the desired effect.

**Comment by niieani** [comment]
@martingust I've tried adding `view-cache="*"` to all three examples that I've posted to no effect, confirming what @stsje said. 

If it should work, it doesn't, so we have probably discovered a bug. In any case, shouldn't this be an **opt-out** feature instead of **opt-in**? Having it opt-in sounds like premature optimization to me - and in many cases will have the opposite, de-optimizing effect. If the developer needed the extra bit of performance and willingly never changes the instance underlying the repeater, he could then choose to **_opt out_** of the cache. I'd imagine most users will want such basic caching / item diffing by default, without digging deep into the docs (which, by the way, currently don't mention anything about `view-cache` or how to use it: it would be great to see a blog article about it at some point!).

**Comment by martingust** [comment]
@niieani @stsje Sorry for misleading about the `view-cache`, it won't make Aurelia not destroy/create views that require life cycle, but it should speed it up by caching the DOM elements. However I think this is something we should look into to optimize more.

**Comment by niieani** [comment]
@martingust Oh, that explains it. I've just done some performance tests with and without `view-cache` setting and there seems to be no difference in performance, even with 1000 or more elements in the repeater. So the view-caching filter does not help with this use-case at all (if so, the differences are in the millisecond range for 1000s of elements).

**Comment by jdanyow** [comment]
Based on @niieani's comments they understand that returning a different array instance from a value converter is expensive because it forces Aurelia to re-render the whole repeat.  Aurelia does not do any expensive DOM diffing computations, nor does it have a copy of the array it can compare the new array instance with to determine what the effective changes were.

Aurelia's approach is to intercept array mutation and make the minimal DOM changes necessary to sync up the DOM. This doesn't work well when a new array instance is returned with each evaluation of a value converter.

I agree the value converter approach @niieani is using is the most convenient *existing* way for developers to implement the toggle filtering on/off scenario.  That said, I doubt there is anything further we can do to support scenarios where the repeat's array instance is replaced.  I don't think we're going to add DOM diffing, there's other frameworks for that, and it tends to be slower and heavier in a lot more use cases.

Maybe what we need to do is make it more natural for developers to express this particular category of  binding expressions:  *dynamically filtered repeats*...  What if we supported something like this:

```html
<li repeat.for="item of items where item.foo > 6 || bar">
```

In other words, extend the repeat expression syntax to allow supplying a predicate that would be used to determine whether an item should be rendered. I think this might solve a whole category of issues people have with repeats. With this kind of syntax we can observe the predicate expression and show/hide items when `item.foo` changes or when `bar` changes.  This is difficult to do today, which is why a lot of people are using hacks like `this.array = this.array.slice(0)` to trigger re-rendering of repeats.

Thoughts?

**Comment by EisenbergEffect** [comment]
@jdanyow Dang. That would be so cool. I love the expressiveness of it. What would be the level of effort in implementing something like this?

@niieani Thoughts on this?

@jdanyow @martingust Regardless of the above, I do think we should continue to look for potential optimizations in repeat scenarios.

**Comment by CasiOo** [comment]
@niieani's Problem could possibly be fixed by chaining filters like `<li repeat.for="item of items | standardFilter | repeatDiffFilter>`. So you have a filter that does the diffing.

@jdanyow Your idea is interesting indeed, but I would personally need a clear defined rule for when to use which kind of filter. Two kinds of filters could become messy. It will have to be very clear to the developer when to use which, and what the pros and cons are. Would like to see more work on it before commenting if it's a good or bad idea to have filter #1 and filter #2.

**Comment by niieani** [comment]
Initially I thought @jdanyow's `where X` syntax was a good idea, but after thinking more about the actual scenarios, I'm now not so optimistic about it.
Here are my reasons:

- this forces people to put more logic into the view
- sometimes the filtering is really complex - this could mean lines and lines of less-readable, non-lintable code inside of HTML templates that additionally needs to be parsed by Aurelia with each load
- we already have ValueConverters, why introduce another concept to learn + non-spec complexity for the parser?

Finally, filtration is only one of the reasons for why one would like to replace the instance of the underlying Array/Map/Set, rather than mutate it. The other huge one is immutability, which is already big, and it's getting bigger in the JS world every day. Just think of possible future integrations with all those new, cool server<->client frameworks that use GraphQL-like data pushing, including [Facebook's Relay](https://facebook.github.io/relay/), [Netflix's Falcor](http://netflix.github.io/falcor/), plus soon-to-come Meteor's new [Apollo stack](https://github.com/apollostack/apollo) or RethinkDB's new [Horizon](https://rethinkdb.com/blog/new-in-2016/) platform.

They all work with immutable data, which means array instances **will change** whenever the remote object is pushed to the client. I should also mention Observables, now on their way to become standardized in ES2017, are often used to generate immutable Arrays (e.g. as often used in Cycle.js).

Mutating arrays is fine for small, "garage" projects, but if Aurelia wants to compete on the market for medium or large enterprise deployments, I think it should support immutable data in such cases (i.e. Array instance replacement).

I don't think Aurelia needs to do DOM diffing though - that would be too much - since it already has the cool mutation handler, as @jdanyow points out:

> Aurelia's approach is to intercept array mutation and make the minimal DOM changes necessary to sync up the DOM

I'd suggest just taking the "interception" idea one step further. Don't only intercept the mutations of the array, but also intercept changes of the instance, compare both instances' values and only then do the mutation accordingly. I don't think it's that complicated actually, and @CasiOo suggested it could even be solved by an additional ValueConverter (`repeatDiffFilter`) that does the diffing (excellent idea!).
Even so, I think this could be baked into Aurelia's core, because:

- if the instance doesn't change, there is no performance impact for the user (the code only runs when the instance _does_ change)
- if it does change, the user should be able to benefit from the increased performance without knowing about and adding any extra code/converters

**Comment by stsje** [comment]
Will the value-converter continue to be a singleton even if you add the `@transient()` decorator?

**Comment by niieani** [comment]
So I've [tried implementing](https://gist.run/?id=499ef4b8b871105deb9fdf597e47265d) a `repeatDiff` value converter, however I've stumbled onto a bug in Aurelia #214. Once that is resolved, **[this proof-of-concept `DiffValueConverter`](https://gist.run/?id=499ef4b8b871105deb9fdf597e47265d)** should work to optimize for changing array instances.

@stsje Didn't know about the `@transient()` decorator, which Aurelia package is it available in for import?

**Comment by stsje** [comment]
Import it from aurelia-framework and add it to your class `import {transient} from 'aurelia-framework';

Seems weird that the value-converter changes to the array aren't picked up by Aurelia. Probably because it's the same instance returned. That was not what I expected to happen.

**Comment by niieani** [comment]
@stsje Indeed, hence #214.

I was just thinking about implications of using such a "caching and diffing" ValueConverter. Would such a `@transient` ValueConverter instance be recreated in this case we signal the binding? Like this:

```js
repeat.for="element of elements | transientConverter & signal: 'some-signal'"
```

If yes, then even a DiffValueConverter don't help us solve the issue entirely, because it'll be recreated every time the binding is signaled, resulting in a new array being created and Aurelia removing and re-adding all the elements again, which is what we're trying to avoid.

**Comment by niieani** [comment]
Another downside to using such an "instance diffing" ValueConverter is that not all optimizations can be made in it. The ValueConverter can only optimize for elements added and removed, but not for elements moving around in the Array. Having this baked into Aurelia could mean that Aurelia could simply shift the DOM elements around in case of element reordering, e.g. when the difference between two arrays is that the new one is sorted. 

Imagine having a table/datagrid, filterable and sortable by multiple fields (e.g. Excel's autofilter) - right now when we replace the array with its sorted version, Aurelia has to unbind and remove all the elements from the DOM, create and bind them and them re-add them to the DOM again. If it did array-element diffing, it could just re-position the already existing DOM elements intelligently, giving huge performance benefits for such sortable/filterable data-grids.
If the `ui-virtualization` plugin suffers from the same problem (is there a way to include it in the gist.run to test?), it could also benefit if such a feature were present in the binding system.

**Comment by jdanyow** [comment]
@niieani - good feedback, thank you.  I have some feedback/questions on your comments.  I think generally we're moving in the right direction.  If you have a few minutes could you respond to my comments below?  Much appreciated!

>I'd suggest just taking the "interception" idea one step further. Don't only intercept the mutations of the array, but also intercept changes of the instance, compare both instances' values and only then do the mutation accordingly.

This would only work for the `this.array = this.array.slice(0)` hack people are using to force value converter re-evaluation right?  If it were a truly immutable data model or a fresh set of data from the server, the array instance would be different from the original instance, *as would the objects that are the content of the array*.  For truly immutable data scenarios we'd need a DOM diff wouldn't we?  If we agree on that then I think the next question would be *how would a DOM diff work with custom elements that are stateful?*  For example if you had `<foo bar.bind="item.obj"></foo>` in a repeat template?

Some questions/feedback on your comments related to the `where` idea:

>* this forces people to put more logic into the view

Is it really *more* logic in the view than what the converter approach takes?  Using the gists from the original post as an example, which one of these is better:

```html
<component repeat.for="id of components | without8 : isFiltering.checked"></component>
<component repeat.for="id of components where !isFiltering.checked || id !== 8"></component>
```

>* sometimes the filtering is really complex - this could mean lines and lines of less-readable, non-lintable code inside of HTML templates that additionally needs to be parsed by Aurelia with each load

This would be an argument against both the value-converter approach and the `where` approach right?  If not, can you give an example?

>* we already have ValueConverters, why introduce another concept to learn + non-spec complexity for the parser?

What do you mean by non-spec?
As for why introduce another concept, in addition to the performance issue you've created here, expressing the filter in the view via the `where` idea solves an often reported issue where people don't understand why changing a property on an object in array doesn't cause the valueconverter/filter to re-execute. Having the filter expressed via the `where` allows the binding system to observe the minimal set of props to achieve the right behavior.

**Comment by niieani** [comment]
> @niieani - good feedback, thank you. (..) Much appreciated!

Happy to help! I want this project to succeed as much as you, you're doing great work!

> This would only work for the this.array = this.array.slice(0) hack people are using to force value converter re-evaluation right? (...) For truly immutable data scenarios we'd need a DOM diff wouldn't we? 

Only if the underlying data actually changes or is pushed again by the server. You can have an immutable list of immutable objects that you're rearranging, mapping to other immutable objects or simply filtering through. In all those cases the underlying elements of the list refer to the same objects in memory and it is **only the list's instance** that changes. This is the case I was referring to in this issue and I think it solves most use cases for repeaters, however you point out the other case I didn't think about before:

> If it were a truly immutable data model or a fresh set of data from the server, the array instance would be different from the original instance, as would the objects that are the content of the array. 

That's a very good point @jdanyow. 
This will happen specifically if a server resends the same data more than once, effectively swapping the array of objects with a new array of entirely new objects.

However, I still think DOM / VDOM diffing might not be necessary. 
Here's a stab at a solution to this particular problem:

- have two "modes" that the repeaters could use to match the objects: 
  - **default** - compare instances of repeater's elements
  - **custom** - the user provides a property uniquely identifying the instance of the object inside the array, sort of like a "hashcode" or a unique ID that the Aurelia can use to swap in the bindingContext of the rendered repeat template

In the second case, if the user doesn't provide such unique property, Aurelia could either resolve to today's behavior (reset/recreate all repeater items) or it could deep-diff the elements. There's [a library](https://github.com/flitbit/diff) for such purpose and it's only 4.4KB minified. I'd imagine diffing the underlying objects directly performance-wise might be as good as VDOM diffing, if not better.

The code would only be triggered when the instance changes, so for a lot of use cases there would not be any performance implications either way.

Something even more flexible than providing a unique property, could be a comparison function. So the user could write custom logic that the repeater would execute to locate the copy of the object among the previous instance's element.

> Is it really more logic in the view than what the converter approach takes?

The examples I provided are really trivial. But how would you convert this filter into the `where x` syntax? (taken from one of my projects currently in production):

```ts
export class FilterSearchValueConverter {
  toView(groups:Array<IGroup>, search:string, decryptedProperty = 'title') {
    if (!groups || groups.length === 0 || !search || search.length < 3) return groups;
    return groups.filter(group => group.decrypted && group.decrypted[decryptedProperty] ? group.decrypted[decryptedProperty].toLocaleLowerCase().search(search.toLocaleLowerCase()) > -1 : false);
  }
}
```

And what about [sorting converters](https://github.com/Vheissu/Aurelia-Code-Snippets/blob/master/Value%20Converters/sort-array.js) or value converters that are mapping array values? `where` syntax is only useful for filtration, but does not address a range of other transformations.

> What do you mean by non-spec?

As I understand, @EisenbergEffect prefers Aurelia's functionality to be as close to JavaScript and HTML specs as possible and that includes binding language, which is very similar to JavaScript (currently with the exception of `|` and `&`). `where` would have to be another reserved keyword and another concept to learn with Aurelia. Not a big deal for me personally, just wanted to point out that a new concept increases complexity.

> expressing the filter in the view via the `where` idea solves an often reported issue where people don't understand why changing a property on an object in array doesn't cause the valueconverter/filter to re-execute. Having the filter expressed via the `where` allows the binding system to observe the minimal set of props to achieve the right behavior.

Indeed. However, to address that problem - wouldn't it be easier to just add something akin to the signal binding behavior that recomputes the binding once a value changes? Something like this:

```html
<component repeat.for="element of components | something & recomputeOnChange: element.date"></component>
```

Hope that helps.

**Comment by EisenbergEffect** [comment]
Great conversation here. Keep it coming. Let's see how we can improve this!

**Comment by niieani** [comment]
Here's an experiment using the [DeepDiff](https://github.com/flitbit/diff) library that swaps the arrays using a proxy array to which we apply the results of the diff of the two. The View then binds to that proxy array, making use of the ArrayObserver:
https://gist.run/?id=90555919452fff29a7abe1440a5d5556

The result is that the amount of removed and added elements is relatively minimal (have a look in the console). I believe this is something which could be implemented in Aurelia directly.

**Comment by EisenbergEffect** [comment]
@niieani Using that technique, are you seeing the sort of performance you need?
@jdanyow @martingust Definitely we should look into this.

**Comment by niieani** [comment]
@EisenbergEffect yes, even with 9000 elements in the array the swap is now instantaneous. Have a look yourself: https://gist.run/?id=a986f5c6e3c5202d64b3d49a084981d0

**Comment by stoffeastrom** [comment]
>Mutating arrays is fine for small, "garage" projects, but if Aurelia wants to compete on the market for medium or large enterprise deployments, I think it should support immutable data in such cases (i.e. Array instance replacement).

Why not use [JSONPatch](https://tools.ietf.org/html/rfc6902)?
For a large enterprise app written in angular 1.5 this was really needed to avoid creating new object references.

I'm not sure I get the point of having 9000 elements in an array and render them :)?

**Comment by niieani** [comment]
@stoffeastrom The only point of 9000 elements is a benchmark. No practical application.

**Comment by stoffeastrom** [comment]
@niieani I'm still trying to understand why you would want the server to resend that many items to achieve "immutable". Imho that seems a bit verbose

**Comment by niieani** [comment]
@stoffeastrom Feel free to read up on the architectures of any of these technologies:

- [Meteor Apollo stack](https://github.com/apollostack/apollo)
- [RethinkDB Horizon platform](https://rethinkdb.com/blog/new-in-2016/)
- [Facebook GraphQL](https://facebook.github.io/graphql/)
- [Facebook Relay](https://facebook.github.io/relay/)
- [Netflix Falcor](http://netflix.github.io/falcor/)

I think it would be best not to pollute the topic with explanations, but focus on the discussion regarding the best implementation of a solution. 

But I'll give you a quintessential example (below) and a brief explanation with a use-case: every object that is represented in the database can be pushed by the server to the client according to the specification of the client. Because objects are "reactive" - i.e. we're listening to any changes made to it, the server can push them at any time to the client. Such changes are pushed as new "immutable" objects. Note that I put _immutable_ in the quotation marks - they really are simple JSON objects that you could modify if you wanted to, and the _"immutability"_ is really only a best practice and not a requirement. However, since we're dealing with such data you'll almost always want to persist your changes in the database, thus you'll want to use your framework's APIs to commit those changes anyway. Hence, it's really not very useful to mutate the objects directly -- enter: "immutability". 

It turns out "verbosity", as you call it, is really not a problem either. As you can see there's a huge shift to these sort of technologies in the client<->server arena among all the big players, simply because it's easier to reason about data when you don't have to deal with object mutations too often. And the performance impact is negligible, even on low-end devices such as phones (the reason is that old data without references is quickly garbage-collected anyway, and in some cases mutations can even be _slower_ than freshly created objects!).

You can imagine a chat application, like Gitter, where every message is an object in the database. You could easily get to a few hundred elements in an array in the client (messages) that represents the state of the open chat. If somebody edits a chat message, a new object with the same ID would be pushed to the client - that object should then be replaced/updated by its client's representation to reflect the remote changes. 

@jdanyow Since in most of the listed server-client frameworks, the way in which new updates are handled is an implementation detail, I believe simple shallow comparison of repeater's elements would be sufficient for at least 90% of use-cases. It solves: filtering, mapping, sorting, adding/removing individual repeater elements; all of which are useful not only for client<->server synchronization frameworks, but also Observables / RX along with similar technologies, and finally solves array ValueConverters. 

The only case I can think of, where deep-comparison of array's elements would be required is if the architecture of the "reactive" object contains its own nested arrays with objects. Even then, it's an anti-pattern to store large arrays inside of objects, so I don't think it would be a problem in too many cases. Can anybody think of any other use-case?

To go back to the example, the only time the server would send out lots of new "immutable" objects is the initial load, and that happens once. After that all the changes are incremental (think of the chat again: added message, removed message, edited message, filtering messages - search). So even if we have to unbind/remove the same object when it's edited version is pushed by the server, creating a new Object reference, it would only be an individual change, so the performance impact of re-rendering would be negligible. 
Unless someone presents a valid use-case for deep-object comparison I think we could focus on the simpler, basic implementation as a first step (like in my experimental concept) as a solution to the majority of use cases and then go from there if a need for more complex solutions is presented.

**Comment by niieani** [comment]
@EisenbergEffect @jdanyow So I've had a go at implementing this. Not so complicated after all.
Two pull requests as I needed to add one simple function to _aurelia-templating_ (moving Views across a ViewSlot):

- [perf(repeat): intelligently reuse items from previous instance](https://github.com/aurelia/templating-resources/pull/215) #215
- [feat(ViewSlot): moving Views across the slot #324](https://github.com/aurelia/templating/pull/324)

Here's a demo:
- https://gist.run/?id=af08bd464f72a809d07639b2d9f88a38

**Comment by EisenbergEffect** [comment]
@niieani Awesome! Thanks for working on this. I've cc's @jdanyow and @martingust on those PRs. They are the experts in that area, so they can review and work with you to make any changes needed.

**Comment by EisenbergEffect** [comment]
@jdanyow For comparing different instance that are actually to be treated as the same, I think we had something we used for selects on that. Maybe we can use the same thing here?

**Comment by stsje** [comment]
@niieani Maybe I have missed it, but how are duplicated values being handled?

**Comment by niieani** [comment]
@stsje See [these two tests](https://github.com/niieani/templating-resources/blob/5ad659b852931da26c79d2b204fc0424aebb37bd/test/repeat-integration.spec.js#L624-L716). In case the new instance contains duplicates, the original View is re-used and additional new views are created for the duplicates. When removing duplicates, the same thing happens in the opposite order (we keep the required amount of duplicate views). This is why when iterating and updating the views with the new items I'm [setting 'undefined' on the cache of old items](https://github.com/niieani/templating-resources/blob/5ad659b852931da26c79d2b204fc0424aebb37bd/src/array-repeat-strategy.js#L84) is required - otherwise `indexOf` would always return the index of the first reference to the item.

**Comment by stsje** [comment]
Good to see you've given it some thought! :+1: 
I imagine it's difficult to make sure duplicate values are being moved the right way (for animation)

**Comment by niieani** [comment]
@stsje Well, if the value is a duplicate there's no way to tell the intention of the change's author - if he meant it as movement or if he meant it as a new value, because that information simply isn't there (arrays  aren't movement vectors _<how cool would that be?!>_). But the key question is - if you do have duplicates, does it really matter which duplicate you move to which place? After all, the View generated from the duplicate is going to be identical to the original. I can't think of any case in which it would make sense to distinguish them, and in those that it does make sense, the objects themselves would be copies (and therefore different references), perhaps with a different ID inside of them.
